src/nbl/asset/utils/CMeshManipulator.cpp

```diff
@@ -389,7 +396,41 @@ core::smart_refctd_ptr<ICPUMeshBuffer> IMeshManipulator::calculateSmoothNormals(
 		return nullptr;
 	}
 
-	auto outbuffer = makeNewMesh ? createMeshBufferDuplicate(inbuffer) : core::smart_refctd_ptr<ICPUMeshBuffer>(inbuffer);
+    core::smart_refctd_ptr<ICPUMeshBuffer> outbuffer;
+    if (makeNewMesh)
+    {
+        outbuffer = core::move_and_static_cast<ICPUMeshBuffer>(inbuffer->clone(0u));
+
+        const auto normalAttr = inbuffer->getNormalAttributeIx();
+        auto normalBinding = inbuffer->getBindingNumForAttribute(normalAttr);
+        const auto oldPipeline = inbuffer->getPipeline();
+        auto vertexParams = oldPipeline->getVertexInputParams();
+        bool notUniqueBinding = false;
+        for (uint16_t attr=0u; attr<SVertexInputParams::MAX_VERTEX_ATTRIB_COUNT; attr++)
+        if (attr!=normalAttr && (vertexParams.enabledAttribFlags&(0x1u<<attr))!=0u && vertexParams.attributes[attr].binding==normalBinding)
+            notUniqueBinding = true;
+        if (notUniqueBinding)
+        {
+            int32_t firstBindingNotUsed = core::findLSB(vertexParams.enabledBindingFlags^0xffffu);
+            assert(firstBindingNotUsed>0 && firstBindingNotUsed<SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT);
+            normalBinding = static_cast<uint32_t>(firstBindingNotUsed);
+
+            vertexParams.attributes[normalAttr].binding = normalBinding;
+            vertexParams.enabledBindingFlags |= 0x1u<<normalBinding;
+        }
+
+        const auto normalFormatBytesize = asset::getTexelOrBlockBytesize(inbuffer->getAttribFormat(normalAttr));
+        auto normalBuf = core::make_smart_refctd_ptr<ICPUBuffer>(normalFormatBytesize*IMeshManipulator::upperBoundVertexID(inbuffer));
+        outbuffer->setVertexBufferBinding({0ull,std::move(normalBuf)},normalBinding);
```

Comment:
dont you have to copy (via getAttribute, setAttribute though) normals from inbuffer?

Reply:
the function for computing smooth normals does `getAttribute` from the source buffer to compute the smooth normal and then does `setAttribute` on all vertices it considers to be part of the smoothgroup so no need to copy the data